apps/ejabberd/src/mod_mam_utils.erl

```diff
@@ -346,6 +346,25 @@ is_valid_message_children(_,     false, false, false) -> true;
 %% See mam_SUITE:offline_message for a test case
 is_valid_message_children(_,      _,    _,     _    ) -> false.
 
+is_chat_marker(Packet) ->
+    case mod_mam_params:archive_chat_markers() of
+        true ->
+            do_check_markers(Packet);
+        false ->
+            false
+    end.
+
+do_check_markers(Packet) ->
+    MarkerTags = [<<"received">>, <<"displayed">>, <<"acknowledged">>],
+    HasSubtag = fun(Tag) ->
+                    case xml:get_subtag(Packet, Tag) of
```

Comment:
1. We tend to use `exml` and `exml_query` recently to inspect packets.
2. There is (in newest exml) `exml_query:path/2` with selector `{element_with_ns, NS}` which you can use to get tags in specified namespace.
3. You need to check the namespace after all to make sure the `<received/>` tag is really a chat marker.